sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -508,7 +508,14 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       sparkSession.sessionState.conf.sessionLocalTimeZone)
     val filteredLines: Dataset[String] =
       CSVUtils.filterCommentAndEmpty(csvDataset, parsedOptions)
-    val maybeFirstLine: Option[String] = filteredLines.take(1).headOption
+    val needsFirstLine: Boolean =
+      userSpecifiedSchema.isEmpty || parsedOptions.headerFlag
+    val maybeFirstLine: Option[String] =
+      if (needsFirstLine) {
+	filteredLines.take(1).headOption 
```

Comment:
The indentation of this is still off. The title on the PR still needs to be updated, with a JIRA. Please see my other comment. You can probably inline `needsFirstLine` too to keep it simpler.

Otherwise I think this is pretty OK, AFAIK, the first line is never needed unless it's a header or needed for schema inference.

Reply:
Sorry, about those issues. They should be resolved now. Thanks!